akka-stream-tests/src/test/scala/akka/stream/scaladsl/FoldWithinSpec.scala

```diff
@@ -0,0 +1,244 @@
+    timePasses(maxGap)
+
+    p.sendNext(3)
+    timePasses(maxGap/2) // less than maxGap should not cause emit
+    p.sendNext(4)
+    timePasses(maxGap)
+
+    p.sendNext(5)
+    timePasses(maxGap/2) // less than maxGap should not cause emit
+    p.sendNext(6)
+    timePasses(maxGap/2) // less than maxGap should not cause emit and does not accumulate
+    p.sendNext(7)
+    p.sendComplete()
+
+    Await.result(result, 10.seconds) should be (Seq(Seq(1, 2), Seq(3, 4), Seq(5, 6, 7)))
+
+  }
+
+  "split aggregator by total duration" in {
+    val maxDuration = 400.seconds
+
+    val p = TestPublisher.probe[Int]()
+
+    val result = Source.fromPublisher(p)
+      .via(
+        new FoldWithin[Int, Seq[Int], Seq[Int]](
+          seed = i => Seq(i),
+          aggregate = (seq, i) => seq :+ i,
+          emitOnAgg = _ => false,
+          harvest = seq => seq,
+          maxDuration = Some(maxDuration), // elements with longer gap will put put to next aggregator
+          getSystemTimeMs = getSystemTimeMs
+        )
+      )
+      .runWith(Sink.collection)
+
+    p.sendNext(1)
+    timePasses(maxDuration/4)
+    p.sendNext(2)
+    timePasses(maxDuration/4)
+    p.sendNext(3)
+    timePasses(maxDuration/4)
+    p.sendNext(4)
+    timePasses(maxDuration/4) // maxDuration will accumulate
+
+    p.sendNext(5)
+    p.sendNext(6)
+    p.sendNext(7)
+    p.sendComplete()
+
+    Await.result(result, 10.seconds) should be (Seq(Seq(1, 2, 3, 4), Seq(5, 6, 7)))
+
+  }
+
+  "down stream back pressure should not miss data on completion without pull on start" in {
+
+    val maxGap = 1.second
+    val upstream = TestPublisher.probe[Int]()
+    val downstream = TestSubscriber.probe[Seq[Int]]()
+    // Note that the cost function set to zero here means the stream will accumulate elements until completed
+    Source.fromPublisher(upstream).via(
+      new FoldWithin[Int, Seq[Int], Seq[Int]](
+        seed = i => Seq(i),
+        aggregate = (seq, i) => seq :+ i,
+        emitOnAgg = _ => false,
+        harvest = seq => seq,
+        maxGap = Some(maxGap), // elements with longer gap will put put to next aggregator
+        getSystemTimeMs = getSystemTimeMs,
+        pullOnStart = false
+      )
+    ).to(Sink.fromSubscriber(downstream)).run()
+
+    downstream.ensureSubscription()
+    upstream.sendNext(1)
+    timePasses(maxGap) // this will not cause emit, without eager pull the data does not make to the next stage until downstream request
+    upstream.sendNext(2)
+    downstream.request(1)
+    downstream.expectNoMessage() // no emit
+    timePasses(maxGap) // this will cause emit
+    downstream.expectNext(Seq(1, 2))
+    upstream.sendNext(3)
+    upstream.sendNext(4)
+    downstream.request(1)
+    downstream.expectNoMessage()
+    upstream.sendComplete() // flush the pending data
+    downstream.expectNext(Seq(3, 4))
+    downstream.expectComplete()
+  }
+
+  "down stream back pressure should not miss data on completion with pull on start" in {
```

Comment:
@johanandren does this test case capture the corner case you have in mind?

Reply:
I thinks so, closing upstream and emitting even though emitOnAgg says false 👍